mlir/lib/Conversion/TosaToRock/TosaToRock.cpp

```diff
@@ -1704,106 +1712,222 @@ struct AttentionRewritePattern : public OpRewritePattern<tosa::MatMulOp> {
     if (shape.size() != 4 || shape[0] != 1 || shape[1] != 1)
       return false;
 
-    // Ensure that element type is integer
-    if (!constAttr.getElementType().isIntOrIndex())
+    // Sanity check that this is an integer when expectOnesInUpperTriangle is
+    // false (i.e., the select case), and a float when expectOnesInUpperTriangle
+    // is true (i.e., the add case).
+    bool isInt = constAttr.getElementType().isIntOrIndex();
+    bool isFloat = isa<FloatType>(constAttr.getElementType());
+    if ((!isInt && !isFloat) || (isInt && !expectOnesInUpperTriangle) ||
+        (isFloat && expectOnesInUpperTriangle))
       return false;
 
     int64_t seqLen = shape[2];
     int64_t maxSeqLen = shape[3];
-    auto values = constAttr.getValues<APInt>();
-    for (int64_t row = 0; row < seqLen; ++row) {
-      for (int64_t col = 0; col < maxSeqLen; ++col) {
-        auto val = values[row * maxSeqLen + col].getSExtValue();
-
-        // Validate that the lower triangular portion is all zeros
-        if (col <= row && val != 0)
-          return false;
-
-        // Check that the rest is all just 0's and 1's
-        if (col > row && val != 1)
-          return false;
+
+    // Generic validation function that works with any value type
+    auto validateMask = [&](auto values, auto isZero, auto isOne,
+                            auto isLargeNegative) -> bool {
+      for (int64_t row = 0; row < seqLen; ++row) {
+        for (int64_t col = 0; col < maxSeqLen; ++col) {
+          auto val = values[row * maxSeqLen + col];
+
+          // Validate that the lower triangular portion is all zeros
+          if (col <= row && !isZero(val))
+            return false;
+
+          // Check that the upper triangular portion is correct
+          bool validUpperTriangleVal =
+              expectOnesInUpperTriangle ? isOne(val) : isLargeNegative(val);
+          if (col > row && !validUpperTriangleVal)
```

Comment:
```suggestion
          if (!validUpperTriangleVal)
```

nit: 
don't really need to check for `col > row`

Reply:
why not?

Reply:
I do think we need this. We only want to be checking  `validUpperTriangleVal`  for the values in the upper triangular portion.

Reply:
It does need it. my bad.